src/interrupt/mod.rs

```diff
@@ -0,0 +1,207 @@
+            Error::InterruptNotChanged => write!(f, "the interrupt state could not be changed"),
+            Error::InterruptNotTriggered => write!(f, "the interrupt could not be triggered"),
+        }
+    }
+}
+
+/// Trait used by interrupt producers to signal interrupts.
+///
+/// An object having the Interrupt Trait is shared between the VMM and the device
+/// that is using the interrupt.
+/// The `Interrupt` performs two main goals:
+///     * offers a control interface for the interrupt through the `enable()`
+///     and `disable()` methods;
+///     * offers a channel through which notifications can be passed between
+///     the VMM and the device;
+///
+/// When an `Interrupt` is triggered, a notification mechanism that is known by the
+/// Hypervisor will be used to signal the guest. The type of the notification mechanism
+/// is defined by the `NotifierType` associated type.
+/// `Interrupt` allows access to the undelying mechanism used by the Hypervisor through
+/// the `notifier()`. This enables some use cases the device may want to bypass the VMM
+/// completely or when the device crate acts only as a control plane and the actual
+/// emulation is implemented in some other component that understands the underlying
+/// mechanism.
+/// A notable example is VFIO that allows a device to register the irqfd so that
+/// interrupts follow a fast path that doesn't require going through the VMM.
+///
+/// Objects implementing this trait are required to have internal mutability.
+pub trait Interrupt {
+    /// The type of the underlying mechanism used for notifications by this interrupt.
+    type NotifierType;
+
+    /// Inject an interrupt from this interrupt source into the guest.
+    fn trigger(&self) -> Result<()>;
+
+    /// Returns an interrupt notifier from this interrupt.
+    ///
+    /// An interrupt notifier allows for external components and processes
+    /// to inject interrupts into a guest through a different interface other
+    /// than `trigger`.
+    fn notifier(&self) -> Option<Self::NotifierType> {
+        None
+    }
+
+    /// Called back when the CPU acknowledges the interrupt.
+    fn acknowledge(&self) -> Result<()> {
+        Err(Error::OperationNotSupported)
+    }
+
+    /// Returns an end-of-interrupt notifier from this interrupt.
+    ///
+    /// An end-of-interrupt notifier allows for external components and processes
+    /// to be notified when a guest acknowledges an interrupt. This can be used
+    /// to resample and inject a level-triggered interrupt, or to mitigate the
+    /// effect of lost timer interrupts.
+    fn ack_notifier(&self) -> Option<Self::NotifierType> {
+        None
+    }
+
+    /// Enable generation of interrupts from this interrupt source.
+    fn enable(&self) -> Result<()> {
+        Err(Error::OperationNotSupported)
+    }
+
+    /// Disable generation of interrupts from this interrupt source.
+    fn disable(&self) -> Result<()> {
```

Comment:
This trait mixes up multiple aspects of an interrupt:

- The part that is seen by a device (trigger)
- A super-trait of Interrupt where the underlying notifier (eventfd) can be extracted (notifier)
- The part that is operated by interrupt controllers (enable, disable, ack?)

These should be separate traits to simplify implementation and re-usability.

Reply:
Yes, I think separating the notifier related functionality would improve the interface at least for the cases where the notifiers are not available.
For enable/disable I think moving them to another trait has diminishing returns. It's true that those methods are related to interrupt control but that can come either from VMM code or from a device crate code (e.g. a PCI crate which handles the parsing of MSI configuration).

Reply:
I've moved the notifier methods to a different trait. Please take a look.

Reply:
> For enable/disable I think moving them to another trait has diminishing returns. It's true that those methods are related to interrupt control but that can come either from VMM code or from a device crate code (e.g. a PCI crate which handles the parsing of MSI configuration).

I see this from the point of what I have to implement for example to unit test a device that toggles interrupts. A simple device would take an interrupt pin as an input, which can be anything that implements the `trigger` method. When I use a mock of this for unit testing with the current trait, I have to implement lots of unrelated things that the device should not even be able to call at all.

So in my mind it's very important to separate the two sides: One set of functions that the device is supposed to use. One set of functions that the interrupt controller is supposed to use. These should be two traits that also have no relation to each other. That you can implement them on the same struct is an implementation detail.

Or to put it a different way: The device should only be able to use interfaces that it is actually supposed to be using. This prevents people who may not fully understand the interrupt logic from using the interfaces the wrong way, which is a big win.

Reply:
Maybe as clarification: We are currently using rust-vmm code for parts of our virtualization stack at Cyberus Technology internally. We are interested in making it more generic and thus a bit more friendly to hypervisors that don't necessarily work like KVM. :)

This is also a reason why I'm a bit concerned about mixing up concerns in trait implementations and keeping it simple, straightforward and without guess work for the implementor of traits, because this has already caused some churn for us when we onboarded with vm-memory.

Reply:
Gotcha. I think it's good to have the separation for the use cases you mentioned. My simplifying assumption was that most implementations would mix both pure emulation functionality and some interrupt control code so it would be more easier to pass Interrupt objects that have more optional functionality in one larger interface.
But I'm definitely not against splitting them up. Here's what I propose (also merging the discussion about level-interrupts below):

- Interrupt traits **used by device emulation** to generate notifications:
```rust
pub trait EdgeInterrupt {
    fn trigger(&self) -> Result<()>;
}

pub trait LevelInterrupt {
    fn assert(&self) -> Result<()>;
    fn de_assert(&self) -> Result<()>;
}

/// Trait for interrupts that can be masked or unmasked.
/// Can be used with either Level or Edge triggered interrupts.
pub trait MaskableInterrupt {
    fn mask(&self) -> Result<()>;

    fn unmask(&self) -> Result<()>;
}
```

- Interrupt control traits used by interrupt controllers and interrupt configuration code:
```rust
pub trait Interrupt {
    /// Enable generation of interrupts from this interrupt source.
    fn enable(&self) -> Result<()>;

    /// Disable generation of interrupts from this interrupt source.
    fn disable(&self) -> Result<()>;
}

pub trait ConfigurableInterrupt: Interrupt {
    /// Type describing the configuration spec of the interrupt.
    type Cfg;

    /// Update configuration of the interrupt.
    fn update(&self, config: &Self::Cfg) -> Result<()>;

    /// Returns the current configuration of the interrupt.
    fn get_config(&self) -> Result<Self::Cfg>;
}

/// Trait implemented by devices through which a resampled interrupt that was cleared
/// can determine if the device still requires service
pub trait InterruptStatusChecker {
    fn is_active(&self) -> bool;
}

/// Trait for Interrupts that can be automatically retriggered when resampled
/// Uses InterruptStatusChecker
pub trait AutoRetriggerInterrupt: Interrupt {
    fn set_status_checker(&self, Arc<dyn InterruptStatusChecker>) -> Result<()>;
}

pub trait InterruptSourceGroup: Send {
    /// [...]
}
```

- Conversion traits to get the underlying notifiers (for the irqfd use-case)
```rust
pub trait AsRefTriggerNotifier {
    type NotifierType;
    fn trigger_notifier(&self) -> &Self::NotifierType;
}

pub trait AsRefResampleNotifier {
    type NotifierType;
    fn resample_notifier(&self) -> &Self::NotifierType;
}
```

@blitz WDYT?

Reply:
This suggestion looks good to me. The only think I'd like to change is the `de_assert()` method. Maybe `unassert()` or `clear()`?

Reply:
I've updated the PR with the changes discussed here. @gsserge @blitz PTAL